0030-simple-classical-representations.md

```diff
@@ -0,0 +1,319 @@
+
+
+## Summary
+
+This proposes an experimental representation for operations on classical bits solely within conditions.
+This is not required to be the final design for classical handling in Terra, it is meant to be something simple that doesn't entangle existing Terra objects too closely, so can be deprecated in favour of a more complete system in the future.
+
+
+## Motivation
+
+- Current hardware is gaining support for runtime operations and subsequent conditions on classical bits
+- IBM dynamic-circuits hardware can do bit-level calculations, but there is no way to represent this in Qiskit to pass it on
+- In Terra, we need more of a feel for what will actually help us represent more complex classical constructs
+
+
+## User Benefit
+
+- Users of IBM dynamic circuits will be able to take advantage of the classical-processing capabilities of those systems from within Qiskit
+- This support is not limited to IBM so may extend further, but we are not aware of other dynamic-circuits hardware accessible through Qiskit
+- For Terra development: this is a "build one to throw away"-type sytem, so we can see in actual usage what works and doesn't for us before we commit to further classical processing.
+
+
+## Design Proposal
+
+The intent of this RFC is to define an exploratory representation for a simple subset of classical bit-level operations, which expand the current state of Terra's `IfElseOp.condition`.
+This representation is not required to be the base for future expansion; it will be permissible to introduce a completely different classical-value type system if that is more appropriate, and deprecate this one.
+To that end, this proposal will not include adding any additional public methods to any existing Terra classes that might conflict with an alternative implementation.
+In particular, this proposal will not consider adding ease-of-use arithmetic methods to `Clbit` or `ClassicalRegister`, nor will it add expressions that can be used as the parameters to gates.
+
+Notes:
+
+- We want to keep a level of interoperability with the OpenQASM 3 description of hybrid quantum–classical programs, to have better cross-ecosystem compatibility.
+  When in doubt, we probably want to run towards the design choices made there.
+- There is no _requirement_ to throw this representation away again.
+  If it works, we can build on it and keep it.
+
+### Representation
+
+`IfElseOp.condition` and `WhileLoopOp.condition` will expand their allowed values to be an "expression" that evaluates to Boolean.
+The allowed expressions will be the following:
+
+- Object:
+  - literal `bool` (`True` or `False`);
+  - literal `uint[_]` (non-negative Python `int`);
+  - runtime `bool` (`Clbit`);
+  - runtime `uint[n]` (`ClassicalRegister(size=n)`).
+- Unary:
+  - `uint[n] -> uint[n]`: bitwise negation
+  - `bool -> bool`: Boolean negation
+- Binary:
+  - `uint[n] * uint[n] -> uint[n]`: `&`, `|`, `^`;
```

Comment:
Should bit shift `<<` and `>>` be allowed as well?

Reply:
For sure, in the future - this is just meant to be an MVP and I didn't want the available expression tree to be huge for a first attempt.

Reply:
I think given that these are already defined in OpenQASM3 (and available on IBM hardware), and open the door to some applications, it would be good to have these in the first MVP (unless there's a subtlety that would make them more complex to implement than those already listed).

Reply:
That's a good point, and agree that's a good approach for now.